### PR TITLE
Flatpak Support

### PR DESCRIPTION
**Pull request by isantop**
Adds support for adding, listing, and removing Flatpak remotes in Repoman.

This is done conditionally depending on present of a Flatpak support library (pyflatpak) which needs to be present to show the new UI. Without this library, the Flatpak tab is hidden.

Once the library is installed, the tab is available and can be interacted with. It allows removing and listing all currently configured flatpak remotes, both user and system level. It also adds user-level remotes, given a name and a URL to the .flatpakrepo file (with input validation to ensure this looks correct).

For testing this, we should test that installing this does not modify the existing experience on a regular 19.10 installation, as well as checking the actual functionality with pyflatpak installed.

Testing:

 - [ ] Updated Repoman continues to function as presently, without showing the flatpak tab, when pyflatpak is not installed
 - [ ] Flatpak tab shows when pyflatpak is installed.
 - [ ] Adding a flatpak remote (e.g. flathub, https://flathub.org/repo/flathub.flatpakrepo) shows it in the list
 - [ ] Apps can be installed from added flatpak remotes
 - [ ] Configured flatpak remotes are shown in the list.
 - [ ] Flatpak remotes can be removed from the system
 - [ ] If a flatpak remote is added from the command line (`flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo`) it is shown in Repoman after restarting the application.
 - [ ] Adding the same flatpak remote twice only adds it once.
 - [ ] If a flatpak remote is removed in Repoman, all applications installed from it are also removed.

### CODE REVIEW

**Review by mmstick — DISMISSED** [review]

**Comment by leviport** [comment]
I got an error after adding this branch and running updates:

```
-bash: flatpak: command not found
dpkg: error processing package pop-default-settings (--configure):
 installed pop-default-settings package post-installation script subprocess returned error exit status 127
```

Full output is [here](https://github.com/pop-os/repoman/files/4042931/pop-default-settings_install_error.txt) in case it is useful


Do we need to include `flatpak` as a dependency?

**Comment by mmstick** [comment]
Yes, we should have it as a dependency of the desktop metapackage

**Comment by mmstick** [comment]
This should do it: https://github.com/pop-os/desktop/tree/flathub_eoan

**Comment by leviport** [comment]
Thank you @mmstick !

**Comment by isantop** [comment]
It should also probably listed as depends on pyflatpak as well, since that does definitely require flatpak

**Comment by isantop** [comment]
https://github.com/pop-os/pyflatpak/issues/2

**Comment by isantop** [comment]
@leviport It looks like flatpak is already a depends of pyflatpak: https://github.com/pop-os/pyflatpak/blob/master-focal/debian/control

I think that the issue you found was from the `default-settings` package, not from repoman

File: repoman/flatpak.py
Comment (leviport) [inline repoman/flatpak.py]: Should these debug messages be removed / commented out before merging? I do see them if I run repoman from the terminal.

**Comment by leviport** [comment]
Somewhat related to the comment I left about debug messages, I've also noticed that every time I open the "add new source" window, close it (whether I've added a source or not), then open it again, the debug messages become duplicated. An example can be seen here:
```
None
2020-01-17 12:11:27,157 repoman.FPAddDialog DEBUG    Validating name: a, cursor at 0
None
2020-01-17 12:11:31,343 repoman.FPAddDialog DEBUG    Validating name: b, cursor at 0
2020-01-17 12:11:31,343 repoman.FPAddDialog DEBUG    Validating name: b, cursor at 0
None
2020-01-17 12:11:35,326 repoman.FPAddDialog DEBUG    Validating name: c, cursor at 0
2020-01-17 12:11:35,326 repoman.FPAddDialog DEBUG    Validating name: c, cursor at 0
2020-01-17 12:11:35,326 repoman.FPAddDialog DEBUG    Validating name: c, cursor at 0
None
2020-01-17 12:11:39,046 repoman.FPAddDialog DEBUG    Validating name: d, cursor at 0
2020-01-17 12:11:39,046 repoman.FPAddDialog DEBUG    Validating name: d, cursor at 0
2020-01-17 12:11:39,046 repoman.FPAddDialog DEBUG    Validating name: d, cursor at 0
2020-01-17 12:11:39,046 repoman.FPAddDialog DEBUG    Validating name: d, cursor at 0
None
2020-01-17 12:11:42,553 repoman.FPAddDialog DEBUG    Validating name: e, cursor at 0
2020-01-17 12:11:42,553 repoman.FPAddDialog DEBUG    Validating name: e, cursor at 0
2020-01-17 12:11:42,553 repoman.FPAddDialog DEBUG    Validating name: e, cursor at 0
2020-01-17 12:11:42,553 repoman.FPAddDialog DEBUG    Validating name: e, cursor at 0
2020-01-17 12:11:42,553 repoman.FPAddDialog DEBUG    Validating name: e, cursor at 0
None
2020-01-17 12:11:46,030 repoman.FPAddDialog DEBUG    Validating name: f, cursor at 0
2020-01-17 12:11:46,030 repoman.FPAddDialog DEBUG    Validating name: f, cursor at 0
2020-01-17 12:11:46,030 repoman.FPAddDialog DEBUG    Validating name: f, cursor at 0
2020-01-17 12:11:46,030 repoman.FPAddDialog DEBUG    Validating name: f, cursor at 0
2020-01-17 12:11:46,030 repoman.FPAddDialog DEBUG    Validating name: f, cursor at 0
2020-01-17 12:11:46,030 repoman.FPAddDialog DEBUG    Validating name: f, cursor at 0
None
2020-01-17 12:11:49,205 repoman.FPAddDialog DEBUG    Validating name: g, cursor at 0
2020-01-17 12:11:49,205 repoman.FPAddDialog DEBUG    Validating name: g, cursor at 0
2020-01-17 12:11:49,205 repoman.FPAddDialog DEBUG    Validating name: g, cursor at 0
2020-01-17 12:11:49,205 repoman.FPAddDialog DEBUG    Validating name: g, cursor at 0
2020-01-17 12:11:49,205 repoman.FPAddDialog DEBUG    Validating name: g, cursor at 0
2020-01-17 12:11:49,205 repoman.FPAddDialog DEBUG    Validating name: g, cursor at 0
2020-01-17 12:11:49,205 repoman.FPAddDialog DEBUG    Validating name: g, cursor at 0
```
This seems to indicate that a thread/process is being spawned each time the "add" dialog opens, but closing the dialog does not kill the thread. I'm sure it'd take a lot of clicking and closing of that dialog before resources became scarce enough to cause a problem, but would it be an easy fix?

**Comment by isantop** [comment]
@leviport I think what's actually happening is that each new dialog is spawning a new logger, but destroying them isn't getting rid of the old ones. 

It's not a separate process, thread, or event loop though, so apart from the mass of log messages, it should be fairly harmless. It would take many months of straight clicking and closing the dialog for this to cause any real problems. I'll look at fixing it properly too, but I don't think this should hold release, as it's really pretty minor (All of them will be destroyed when the application terminates anyway).

**Comment by leviport** [comment]
So getting rid of the debug loggers will fix the issue? Sounds like a straightforward fix if so.

Another thing I noticed that feels a bit odd: if I double click an entry in the flatpak sources list, it prompts me about whether or not I want to remove it. It seems like the default would be editing the item, even though I don't see a strong reason to need to edit them. At the very least, it's inconsistent with how the list in "Extra Sources" works, which has an "add" button and an "edit" button. Either way, probably something to think about.

File: repoman/flatpak.py
Comment (isantop) [reply repoman/flatpak.py]: It looks like I was initializing the loggers incorrectly, which was causing the duplication. Fixed in 6702852a44d5dde6de83876130bc8486193e6959

**Comment by isantop** [comment]
> Somewhat related to the comment I left about debug messages, I've also noticed that every time I open the "add new source" window, close it (whether I've added a source or not), then open it again, the debug messages become duplicated. An example can be seen here:
> 
> This seems to indicate that a thread/process is being spawned each time the "add" dialog opens, but closing the dialog does not kill the thread. I'm sure it'd take a lot of clicking and closing of that dialog before resources became scarce enough to cause a problem, but would it be an easy fix?

Removing the extra initialization in 6702852a44d5dde6de83876130bc8486193e6959 seems to have also fixed this issue.

**Comment by isantop** [comment]
> So getting rid of the debug loggers will fix the issue? Sounds like a straightforward fix if so.
> 
> Another thing I noticed that feels a bit odd: if I double click an entry in the flatpak sources list, it prompts me about whether or not I want to remove it. It seems like the default would be editing the item, even though I don't see a strong reason to need to edit them. At the very least, it's inconsistent with how the list in "Extra Sources" works, which has an "add" button and an "edit" button. Either way, probably something to think about.

Addressed in c8aee3cc94ea0068d5b7a54713598804c5b9fb44

After talking with @maria-komarova , we're going to remove the double-click handlers in both the Flatpak page and the Extra Sources page. This will help ensure that both of these pages function identically and preserves the consistency of the UX.

**Comment by brs17** [comment]
> 
> Addressed in [c8aee3c](https://github.com/pop-os/repoman/commit/c8aee3cc94ea0068d5b7a54713598804c5b9fb44)
> 
> After talking with @maria-komarova , we're going to remove the double-click handlers in both the Flatpak page and the Extra Sources page. This will help ensure that both of these pages function identically and preserves the consistency of the UX.

For what it is worth, changes in repoman outside of adding flatpak support (such as these changes) should have probably been a separate pr. This would also allow such changes to propagate to 18.04 and 19.10. We'll leave this alone for this pr, but something to think about moving forward.

File: repoman/flatpak.py
Comment (leviport) [inline repoman/flatpak.py]: I think this line has to go as well. I'm still getting `None` in the console for every keystroke in the Name field.

Comment (leviport) [reply repoman/flatpak.py]: Line 106 that is. I'm not sure why it looks like I'm highlighting 3 previous lines as well.

Comment (isantop) [reply repoman/flatpak.py]: Good catch, thanks

**Comment by isantop** [comment]
> For what it is worth, changes in repoman outside of adding flatpak support (such as these changes) should have probably been a separate pr. This would also allow such changes to propagate to 18.04 and 19.10. We'll leave this alone for this pr, but something to think about moving forward.

This was directly related to adding Flatpak support. Otherwise we have UX issues related to inconsistency between the list widgets on both pages, which is not ideal.

**Comment by isantop** [comment]
@leviport We should now be throwing error dialogs for flatpak issues as of 51289ee

**Comment by leviport** [comment]
Nice! So last thing is revoke emoji privileges in text inputs, I believe?

**Comment by isantop** [comment]
@leviport So, it turns out I was wrong and disabling the emoji picker isn't (easily) possible. I can possibly filter those out, but that wouldn't prevent the picker from showing and allowing the user to select one. I can easily hide/show the emoji-picker button, but not enable or disable the picker as a whole.

I will look into possibly filtering emoji from the name, however that might take a bit. I would say that since that functionality isn't very easily exposed (and equally affects the PPA input on the extra sources tab), this shouldn't block merging this branch and should be fixed in a future PR.

**Comment by leviport** [comment]
When adding a flatpak source, I noticed that having a space before the URL makes it fail. Could we trim whitespace off that field before using it?

The error dialog is back now :tada: .   I noticed a couple things that I had questions about.
repoman-error-dialog

Firstly, the messaging it gives upon a failure isn't very informative. It tells me that I got a failure, which is good, but it doesn't tell me why. 

Second, I'm seeing a "Close" button on the right side of the window bar, right next to the "X" button which does the same thing.

**Comment by mmstick** [comment]
It is permitted to have emoji characters in URLs, and there are some "emoji domains" you can get, so perhaps allow them in text entries for URLs?

**Comment by leviport** [comment]
@mmstick I believe we are allowing them in the URL field. The Name field just didn't like them and led to Flatpak giving us errors. Now that we're removing the Name field anyway, we shouldn't have to worry at all about emojis.

**Review by leviport — APPROVED** [review]
Approved for merging into Focal (Please do not merge into Eoan).

**Review by mmstick — APPROVED** [review]

**Comment by isantop** [comment]
@leviport This branch was developed with the intent of not requiring separate codebases for older versions pre-flatpak. It would be highly desirable not to split the branches for this repo and allow merging this into the main `master` branch. Can we figure out some testing whereby that could be a possibility?

**Review by leviport — APPROVED** [review]
Ah yes, I forgot you wrote `pyflatpak` and we could control when that was released. I tested Repoman without `pyflatpak` installed and it worked as expected, so this is good to merge into Eoan and Bionic as well. Just double-check that we won't have any broken dependencies because of the lack of `python3-pyflatpak` available.

**Pull request merged**
